File: packages/i18nify-go/modules/masking/mask_phone_number.go
package masking

import (
	"fmt"
	"strings"
	"unicode"
)

// MaskPhoneOptions configures MaskPhoneNumber behaviour.
type MaskPhoneOptions struct {
	MaskChar     string
	VisibleCount int
}

// MaskPhoneNumber masks all but the last VisibleCount digits of a phone number.
// A leading '+' (international prefix indicator) is preserved.
// Non-digit formatting characters (spaces, hyphens, dots, parentheses) are
// preserved in their original positions.
func MaskPhoneNumber(phone string, opts *MaskPhoneOptions) (string, error) {
	if strings.TrimSpace(phone) == "" {
		return "", fmt.Errorf("masking: MaskPhoneNumber: phone must be a non-empty string")
	}

	maskChar := defaultMaskChar
	visibleCount := defaultVisibleDigits

	if opts != nil {
		if opts.MaskChar != "" {
			maskChar = opts.MaskChar
		}
		if opts.VisibleCount > 0 {
			visibleCount = opts.VisibleCount
		}
	}

	trimmed := strings.TrimSpace(phone)
	hasPlus := strings.HasPrefix(trimmed, "+")
	rest := trimmed
	if hasPlus {
		rest = trimmed[1:]
	}

	digitCount := 0
	for _, ch := range rest {
		if unicode.IsDigit(ch) {
			digitCount++
		}
	}

	maskFrom := digitCount - visibleCount
	if maskFrom < 0 {
		maskFrom = 0
	}

	digitsSeen := 0
	var buf strings.Builder
	for _, ch := range rest {
		if unicode.IsDigit(ch) {
			if digitsSeen < maskFrom {
				buf.WriteString(maskChar)
			} else {
				buf.WriteRune(ch)
			}
			digitsSeen++
		} else {
			buf.WriteRune(ch)
		}
	}

	if hasPlus {
		return "+" + buf.String(), nil
	}
	return buf.String(), nil
}

File: packages/i18nify-js/src/modules/phoneNumber/maskPhoneNumber.ts
import { withErrorBoundary } from '../../common/errorBoundary';
import { DEFAULT_MASK_CHAR, DEFAULT_VISIBLE_DIGITS } from './constants';
import type { MaskPhoneOptions } from './types';

const maskPhoneNumber = (phone: string, options?: MaskPhoneOptions): string => {
  if (!phone || typeof phone !== 'string') {
    throw new Error('maskPhoneNumber: phone must be a non-empty string.');
  }

  const maskChar = options?.maskChar ?? DEFAULT_MASK_CHAR;
  const visibleCount = options?.visibleCount ?? DEFAULT_VISIBLE_DIGITS;

  const trimmed = phone.trim();
  // Preserve leading '+' (international prefix indicator)
  const hasPlus = trimmed.startsWith('+');
  const rest = hasPlus ? trimmed.slice(1) : trimmed;

  // Count total digits (excluding '+')
  const digitCount = (rest.match(/\d/g) ?? []).length;
  const maskFrom = Math.max(0, digitCount - visibleCount);

  let digitsSeen = 0;
  const masked = rest
    .split('')
    .map((ch) => {
      if (/\d/.test(ch)) {
        return digitsSeen++ < maskFrom ? maskChar : ch;
      }
      return ch; // preserve separators (spaces, hyphens, dots, parens, etc.)
    })
    .join('');

  return (hasPlus ? '+' : '') + masked;
};

export default withErrorBoundary<typeof maskPhoneNumber>(maskPhoneNumber);
